File: sdk/python/src/ks_app/mcpproto/legacy.py
"""旧版 MCP 端点（过渡兼容）。

Deprecated: 这些端点使用自定义 JSON 协议，已被标准 MCP Streamable HTTP
端点替代。保留仅用于过渡兼容，将在 Keystone 客户端全部迁移完成后移除。
"""

import logging
from typing import Any, Callable, Awaitable

from starlette.requests import Request
from starlette.responses import JSONResponse
from starlette.routing import Route

from ..context import _reset_meta, _set_meta

_logger = logging.getLogger("ks_app.mcp")


def legacy_call_route(
    tools: dict[str, dict],
    call_tool: Callable[[str, dict], Awaitable[Any]] | None = None,
) -> Route:
    """构造 POST /mcp/tools/call 旧版路由。

    call_tool 回调由上层提供。如果未提供，直接调用 tools[name]["handler"]。
    """

    async def handle(request: Request) -> JSONResponse:
        try:
            body = await request.json()
        except Exception as e:
            return JSONResponse({"error": f"invalid json: {e}"}, status_code=400)

        name = body.get("name")
        params = body.get("params", {})
        tool = tools.get(name)
        if not tool:
            return JSONResponse({"error": f"tool {name} not found"}, status_code=404)

        # 提取 _meta 并注入上下文
        meta = params.pop("_meta", None) if isinstance(params, dict) else None
        _set_meta(meta)
        try:
            if call_tool is not None:
                result = await call_tool(name, params)
            else:
                result = await tool["handler"](**params)
            return JSONResponse({"result": result})
        except Exception:
            _logger.error(
                "tool handler 执行失败",
                extra={"tool": name},
                exc_info=True,
            )
            return JSONResponse({"error": "工具执行失败"}, status_code=500)
        finally:
            _reset_meta()

    return Route("/mcp/tools/call", handle, methods=["POST"])


def legacy_list_route(tools: dict[str, dict]) -> Route:
    """构造 GET /mcp/tools/list 旧版路由。"""

    async def handle(request: Request) -> JSONResponse:
        tool_list = [
            {"name": name, "description": info["description"]}
            for name, info in tools.items()
        ]
        return JSONResponse({"tools": tool_list})

    return Route("/mcp/tools/list", handle, methods=["GET"])

File: sdk/go/mcpproto/legacy.go
package mcpproto

import (
	"encoding/json"
	"log/slog"
	"net/http"
)

// NewLegacyListHandler 创建旧版 GET /mcp/tools/list 端点的 http.Handler。
//
// Deprecated: 此端点是早期的自定义 JSON 协议，已被 NewStreamableHTTPHandler
// 实现的标准 MCP Streamable HTTP 端点替代。保留仅用于过渡兼容，
// 将在 Keystone 客户端全部迁移完成后移除。
func NewLegacyListHandler(tools []ToolDef) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		// 与 streamable handleToolsList 保持一致：annotations 透传到 list 响应。
		// 不强行加 inputSchema 字段以避免 legacy 端点协议形态变化（既有客户端
		// 只期望 name/description）。
		toolList := make([]map[string]any, len(tools))
		for i, t := range tools {
			item := map[string]any{
				"name":        t.Name,
				"description": t.Description,
			}
			if len(t.Annotations) > 0 {
				item["annotations"] = t.Annotations
			}
			toolList[i] = item
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"tools": toolList})
	})
}

// NewLegacyCallHandler 创建旧版 POST /mcp/tools/call 端点的 http.Handler。
// callTool 回调由上层提供，负责执行具体工具逻辑。
//
// Deprecated: 同 NewLegacyListHandler。
func NewLegacyCallHandler(tools []ToolDef, callTool CallToolFunc) http.Handler {
	toolNames := make(map[string]struct{}, len(tools))
	for _, t := range tools {
		toolNames[t.Name] = struct{}{}
	}

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		var req struct {
			Name   string         `json:"name"`
			Params map[string]any `json:"params"`
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			w.WriteHeader(http.StatusBadRequest)
			_ = json.NewEncoder(w).Encode(map[string]any{"error": "请求体解析失败: " + err.Error()})
			return
		}

		if _, ok := toolNames[req.Name]; !ok {
			w.WriteHeader(http.StatusNotFound)
			_ = json.NewEncoder(w).Encode(map[string]any{"error": "tool not found"})
			return
		}

		meta, _ := req.Params["_meta"].(map[string]any)
		delete(req.Params, "_meta")
		ctx := WithMeta(r.Context(), meta)
		result, err := callTool(ctx, req.Name, req.Params)
		if err != nil {
			// handler 返回的 error 可能含内部路径、SQL 片段等敏感信息，不能直接回给客户端。
			// 与新端点 handleToolsCall 保持一致：完整错误通过 slog 记录到服务端日志，
			// 客户端只收到固定提示。
			slog.Error("tool handler 执行失败（legacy endpoint）",
				"tool", req.Name,
				"error", err)
			w.WriteHeader(http.StatusInternalServerError)
			_ = json.NewEncoder(w).Encode(map[string]any{"error": "工具执行失败"})
			return
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"result": result})
	})
}
